src/machine/usb_nxpmk66f18.go

```diff
@@ -0,0 +1,626 @@
+		return
+	}
+
+	// write to second bank
+	data = sendUSB0EP0Chunk(data)
+	if len(data) == 0 {
+		return
+	}
+
+	// save remainder
+	ep0_tx_remainder = data
+}
+
+func sendZlp() {
+	sendUSB0EP0(nil)
+}
+
+func sendUSBPacket(ep uint16, data []byte) {
+	if ep != 0 {
+		panic("not ready to do that")
+	}
+
+	sendUSB0EP0(data)
+}
+
+func (u *USB) CDC() USBCDC {
+	return USBCDC{&u.CDCRXBuffer}
+}
+
+type USBCDC struct {
+	Buffer *RingBuffer
+}
+
+func (u USBCDC) WriteByte(byte) {
+	panic("not implemented")
+}
+
+const (
+	usb_MaxEndpointCount = 16
+	usb_MaxBDTSize       = usb_MaxEndpointCount * 2 * 2
+
+	usb_BDT_Own   = 0x80
+	usb_BDT_Data1 = 0x40
+	usb_BDT_Keep  = 0x20
+	nsb_BDT_NoInc = 0x10
+	usb_BDT_DTS   = 0x08
+	usb_BDT_Stall = 0x04
+
+	usb_BDT_TOK_PID_Mask = 0b00111100
+	usb_BDT_TOK_PID_Pos  = 2
+
+	usb_TOK_PID_OUT   = 0x1
+	usb_TOK_PID_IN    = 0x9
+	usb_TOK_PID_SETUP = 0xD
+)
+
+type register8Aligned32 struct {
+	volatile.Register8
+	_ [3]byte
+}
+
+type usbBufferDescriptor struct {
+	descriptor volatile.Register32
+	address    volatile.Register32
+}
+
+type usbBufferDescriptorTable_t struct {
+	records *[usb_MaxBDTSize]usbBufferDescriptor
+	retain  [usb_MaxBDTSize][]byte
+}
+
+type usbBufferDescriptorEntry struct {
+	record *usbBufferDescriptor
+	retain *[]byte
+}
+
+type usbBufferDescriptorOwner bool
+
+const usbBufferOwnedByUSBFS usbBufferDescriptorOwner = true
+const usbBufferOwnedByProcessor usbBufferDescriptorOwner = false
+
+func (bdt *usbBufferDescriptorTable_t) Address() uintptr {
+	addr := uintptr(unsafe.Pointer(&bdt.records[0]))
+	if addr&0x1FF != 0 {
+		panic("USB Buffer Descriptor Table is not 512-byte aligned")
+	}
+
+	return addr
+}
+
+func (bdt *usbBufferDescriptorTable_t) Entry(i int) usbBufferDescriptorEntry {
+	return usbBufferDescriptorEntry{&bdt.records[i], &bdt.retain[i]}
+}
+
+func (bdt *usbBufferDescriptorTable_t) Get(ep uint16, tx, odd bool) usbBufferDescriptorEntry {
+	i := ep << 2
+	if tx {
+		i |= 2
+	}
+	if odd {
+		i |= 1
+	}
+	return bdt.Entry(int(i))
+}
+
+func (bdt *usbBufferDescriptorTable_t) GetStat(stat uint8) usbBufferDescriptorEntry {
+	return bdt.Entry(int(stat >> 2))
+}
+
+func (bde usbBufferDescriptorEntry) Reset() {
+	bde.record.descriptor.Set(0)
+	bde.record.address.Set(0)
+	*bde.retain = nil
+}
+
+func (bde usbBufferDescriptorEntry) Describe(size uint32, data1 bool) {
+	if size >= 1024 {
+		panic("USB Buffer Descriptor overflow")
+	}
+
+	d := size<<16 | usb_BDT_Own | usb_BDT_DTS
+	if data1 {
+		d |= usb_BDT_Data1
+	}
+	bde.record.descriptor.Set(d)
+}
+
+func (bde usbBufferDescriptorEntry) Set(data []byte, data1 bool) {
+	if len(data) == 0 {
+		// send an empty response
+		*bde.retain = nil
+		bde.record.address.Set(0)
+		bde.Describe(0, data1)
+		return
+	}
+
+	if len(data) >= 1024 {
+		panic("USB Buffer Descriptor overflow")
+	}
+
+	slice := (*reflect.SliceHeader)(unsafe.Pointer(&data))
+
+	*bde.retain = data
+	bde.record.address.Set(uint32(slice.Data))
+	bde.Describe(uint32(slice.Len), data1)
```

Comment:
The reflect package is a relatively large dependency and shouldn't be used in the machine package. It is also possible to do without in this case:

```suggestion
	*bde.retain = data
	bde.record.address.Set(uint32(uintptr(unsafe.Pointer(&data[0]))))
	bde.Describe(len(data), data1)
```